codegen-client/src/main/kotlin/software/amazon/smithy/rust/codegen/client/smithy/customize/SerdeDecorator.kt

```diff
@@ -0,0 +1,28 @@
+/*
+* Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+* SPDX-License-Identifier: Apache-2.0
+*/
+
+package software.amazon.smithy.rust.codegen.client.smithy.customize
+
+import software.amazon.smithy.rust.codegen.client.smithy.ClientCodegenContext
+import software.amazon.smithy.rust.codegen.core.rustlang.Feature
+import software.amazon.smithy.rust.codegen.core.smithy.RustCrate
+
+/**
+ * This class,
+ * - Adds serde as a dependency
+ *
+ */
+class SerdeDecorator : ClientCodegenDecorator {
+    override val name: String = "SerdeDecorator"
+    override val order: Byte = -1
+
+    override fun extras(codegenContext: ClientCodegenContext, rustCrate: RustCrate) {
+        fun _feature(feature_name: String, crate_name: String): Feature {
+            return Feature(feature_name, false, listOf(crate_name + "/" + feature_name))
+        }
+        rustCrate.mergeFeature(_feature("serde-serialize", "aws-smithy-types"))
```

Comment:
Is it possible at all to have a compiler warning if these features are enabled without `--cfg aws_sdk_ustable`?

Reply:
I think you can have a proc macro for it. let me figure it out.

Reply:
What do you think about this?

https://github.com/thomas-k-cameron/compilation-guard

Reply:
Code looks like this
```rust
#![allow(unused_imports, dead_code)]

use compilation_guard::compilation_guard;
#[cfg_attr(all(unstable, feature = "enable-macro"), derive(VeryExpensiveMacro))]
#[cfg_attr(all(not(unstable), feature = "enable-macro"), compilation_guard("lol"))]
struct A;

fn main() {

}
```

and it will result in this if you don't enable the appropriate flag.
```
error: custom attribute panicked
 --> src/main.rs:5:58
  |
5 | #[cfg_attr(all(not(unstable), feature = "enable-macro"), compilation_guard("lol"))]
  |                                                          ^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: message: Compilation guard was triggered!
          lol
```

Reply:
The only downside is that you need to hard code the error message for every attribute, but since the sdk is code-gen-ed I don't think it's a problem.

Reply:
By the way, these features are implemented and pushed here https://github.com/awslabs/smithy-rs/pull/2615

Reply:
Here is an example of generated code.
```rust
// Code generated by software.amazon.smithy.rust.codegen.smithy-rs. DO NOT EDIT.
#[allow(missing_docs)] // documentation missing in model
#[cfg_attr(all(aws_sdk_unstable, feature = "serde-serialize"), derive(::serde::Serialize))]
#[cfg_attr(
    all(not(aws_sdk_unstable), feature = "serde-serialize"),
    ::compilation_guard::compilation_guard(
        "\n
            You must pass \"aws_sdk_unstable\" flag to RUSTFLAG.
            e.g.
            ```bash
            export RUSTFLAG=\"--cfg aws_sdk_unstable\"
            ```
            Learn more about this on this SDK's document."
    )
)]
#[cfg_attr(all(aws_sdk_unstable, feature = "serde-deserialize"), derive(::serde::Deserialize))]
#[cfg_attr(
    all(not(aws_sdk_unstable), feature = "serde-deserialize"),
    ::compilation_guard::compilation_guard(
        "\n
            You must pass \"aws_sdk_unstable\" flag to RUSTFLAG.
            e.g.
            ```bash
            export RUSTFLAG=\"--cfg aws_sdk_unstable\"
            ```
            Learn more about this on this SDK's document."
    )
)]
#[non_exhaustive]
#[derive(::std::clone::Clone, ::std::cmp::PartialEq)]
pub struct CompleteMultipartUploadInput {
```